Improve gRPC patchfile handling

ffoulkes opened the pull request:
- Define a "patchfile" property in the components file. Update components.py to support it.

- Provide three ways to determine the gRPC patch file:

  - Explicitly specified by the `GRPC_PATCHFILE` cmake variable (per the "patchfile" component property). This allows the patch file to be independent of the gRPC version number.

  - Implicitly derived from the `GRPC_VERSION` cmake variable (per the "version" component property). This provides backward compatibility with the current mechanism.

  - Implicitly specified by a `grpc-current.patch.in` symlink in the `patches` folder. This allows an outside process to apply the patch without knowing the gRPC version.

- Refactor patchfile handling into two functions:

  - `define_grpc_patch_file()`, which selects the template file to use.
  - `define_grpc_patch_clause()`, which generates the patchfile from the template and defines the patch clause.

- Make the path to the patch file explicitly relative to the top-level directory (`CMAKE_SOURCE_DIR`) rather than implicitly relative to `grpc.cmake`. This mitigates the "patch file not found" error in the MEV build.

- If we are unable to apply to apply the gRPC patch, the worst outcome is that the RUNPATH of the gRPC plugins won't be updated, and the user will have to set LD_LIBRARY_PATH. This is (a) suboptimal rather than fatal, and (b) moot because this version of the Stratum dependencies breaks RPATH anyway. Make the messages WARNINGs rather than FATAL_ERRORs.

- Correct the version number in the header comment of the gRPC v1.59.2 patch file.

ffoulkes committed:
Overhaul gRPC patchfile handling

- Define a "patchfile" property in the components file.
  Update components.py to support it.

- Provide three ways to determine the gRPC patch file:

  1. Explicitly specified by the GRPC_PATCHFILE cmake variable
     (per the "patchfile" component property). This allows the
     patchfile to be independent of the gRPC version number.

  2. Implicitly derived from the GRPC_VERSION cmake variable
     (per the "version" component property). This provides
     backward compatibility with the current mechanism.

  3. Implicitly specified by a grpc-current.patch.in` symlink
     in the `patches` folder. This allows an outside process to
     apply the current patch without knowing the gRPC version.

- Make a failure to find the patch file a non-fatal error.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

5abeel commented on cmake/grpc.cmake:
```diff
@@ -11,39 +11,115 @@ unset(_grpc_cxx_standard)
+  # Specified patch file
+  if(DEFINED GRPC_PATCHFILE AND NOT GRPC_PATCHFILE STREQUAL "")
+    set(patchfile "${GRPC_PATCHFILE}")
+    if(EXISTS ${PATCHDIR}/${patchfile})
```
Is the user allowed to specify a a ${patchfile} as a build parameter (not sure if that is the intention based on the choice no 1 in comment above)? If yes, adding a ${PATCHDIR} to a provided absolute path will not work

ffoulkes replied:
No. `patchfile` is a local variable.

The semantics of the `GRPC_PATCHFILE` variable are explained at some length in the CL Description. It conveys the value of the new `patchfile` component attribute, just as `GRPC_VERSION` conveys the value of the `version` attribute. I implemented it to provide a flexible alternative to the current naming scheme.

`GRPC_PATCHFILE` _can_ be overridden from the command line (or the preconfig file), as long as you put your alternative patchfile in the `cmake/patches` directory. I don't think that's too great a restriction. And this is _not_ how the variable is intended to be used.

5abeel commented on cmake/grpc.cmake:
```diff
@@ -11,39 +11,115 @@ unset(_grpc_cxx_standard)
-  if(NOT EXISTS ${patchdir}/${patchfile})
-    message(WARNING "Patch file for gRPC version ${GRPC_VERSION} not found")
+  # Current patch file
+  set(patchfile "grpc-current.patch.in")
```
Is there a risk of having a patch file called `grpc-current` and being leftover in a user's build system after a grpc upgrade, resulting in patching mismatch between grpc version and patch file that should be used?

ffoulkes replied:
The semantics of the `grpc-current.patch.in` symlink are explained in the CL Description.

I implemented this feature to provide a fallback mechanism that does not depend on `GRPC_VERSION` or `GRPC_PATCHFILE`.

I created PRs https://github.com/ipdk-io/stratum-deps/pull/46, https://github.com/ipdk-io/stratum-deps/pull/47, and https://github.com/ipdk-io/stratum-deps/pull/48 when you and I were trying to integrate stratum-deps into the MEV build, and were having problems with the patching mechanism.

- PR 46 was a potential fix for the immediate problem. It also made a missing patchfile a warning instead of a fatal error.
- PR 47 provides more control over the definition of the patchfile and continues the move toward using the component manifest to control conditional build behaviors.
- PR 48 implements an experimental mechanism that allows patching to be triggered as a standalone action. I thought this might be useful, given that `repo` downloads stratum-deps without patching it.

Yes, a stratum-deps maintainer could forget to update the symlink when updating gRPC. This would a bug in the update. It's a non-issue until we actually define the symlink and check it in. After that, if someone updates gRPC and forgets to update the symlink, they will presumably discover the error when the build fails during testing.

5abeel reviewed (approved):
LGTM

nupuruttarwar commented on cmake/grpc.cmake:
```diff
@@ -11,39 +11,115 @@ unset(_grpc_cxx_standard)
+
+  define_grpc_patch_file("${patch_dir}" patch_file)
+  if(patch_file STREQUAL "")
+    return()
```
Should we set OUTVAR to empty string before returning here - set(${OUTVAR} "" PARENT_SCOPE)?

ffoulkes replied:
Fair question. I can't remember if I had a specific reason for wanting the value to remain unset. (The comment at lines 84-85 suggests otherwise.) It has the same effect when the ${_patch_clause} substitution is performed.

ffoulkes replied:
Fixed.

nupuruttarwar reviewed (commented):
Do we plan to update the scripts for building dependencies to include the option of specifying GRPC_PATCHFILE ? Also, add documentation that it should be available in cmake/patches directory and on how to do it when building using cmake cmdline?

ffoulkes commented:
> Do we plan to update the scripts for building dependencies to include the option of specifying GRPC_PATCHFILE ? Also, add documentation that it should be available in cmake/patches directory and on how to do it when building using cmake cmdline?

No. GRPC_PATCHFILE is a static mechanism, not a dynamic mechanism. The maintainer defines the `patchfile` component property in `components.json` and then uses the `components.py` script to update `deps.cmake`. The patch step then uses the specified filename instead of deriving one from the gRPC version.

As I noted  in my response to Sabeel, this variable is _not_ intended to be set from the command line. It's a component property, like `version` (which cmake references as GRPC_VERSION).

ffoulkes committed:
Return empty string if patch clause undefined

- Modify define_grpc_patch_clause() to return an empty string if
  define_grpc_patch_file() returns an empty string.

- Improve internal documentation.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes commented:
> Also, add documentation that it should be available in cmake/patches directory and on how to do it when building using cmake cmdline?

@nupuruttarwar - Rereading your comment, I realized that I didn't really address your second question.

The existing documentation is in [How to create a gRPC patch file](https://github.com/ipdk-io/stratum-deps/blob/main/docs/how-to-create-a-grpc-patch-file.md).

To define `GRPC_PATCHFILE`, you add add a `defines` clause to the `grpc` entry in `components.json`:

```json
    "grpc": {
        "name": "grpc",
        "url": "https://github.com/google/grpc.git",
        "sha": "883e5f76976b86afee87415dc67bde58d9b295a4",
        "tag": "v1.59.2",
        "version": "1.59.2",
        "defines": {
            "PATCHFILE": "custom.patch"
        }
    },
```

Then run `components.py` to update the `deps.cmake` file:

```bash
./scripts/components.py -o cmake/deps.cmake
```

Which produces the following:

```cmake
set(GRPC_GIT_URL "https://github.com/google/grpc.git")
set(GRPC_GIT_TAG "883e5f76976b86afee87415dc67bde58d9b295a4") # v1.59.2
set(GRPC_VERSION "1.59.2")
set(GRPC_PATCHFILE "custom.patch")
```

I implemented the `defines` clause as general replacement for `RECENT_PKGS` and other ad hoc ways of handling special cases (especially in Protobuf, which has always been a problem child). See `protobuf.cmake` for examples of past and current workarounds.

We need documentation on `components.json` and the generator script. When time permits...

ffoulkes committed:
Merge branch 'main' into grpc-patch-property
